pkg/azure/blob.go

```diff
@@ -0,0 +1,257 @@
+	for items.NextPage(ctx) {
+		resp := items.PageResponse()
+
+		for _, blob := range resp.ContainerListBlobFlatSegmentResult.Segment.BlobItems {
+			if err := visit(*blob.Name, fmt.Sprintf("%x", *blob.Properties.Etag)); err != nil {
+				err = fmt.Errorf("listing objects from bucket '%s' failed: %w", bucketName, err)
+				return err
+			}
+		}
+	}
+	if err := items.Err(); err != nil {
+		err = fmt.Errorf("listing objects from bucket '%s' failed: %w", bucketName, err)
+		return err
+	}
+	return nil
+}
+
+func (c *BlobClient) Close(_ context.Context) {
+	return
+}
+
+// ObjectIsNotFound checks if the error provided is an azblob.StorageError with
+// an azblob.StorageErrorCodeBlobNotFound error code.
+func (c *BlobClient) ObjectIsNotFound(err error) bool {
+	var stgErr *azblob.StorageError
+	if errors.As(err, &stgErr) {
+		if stgErr.ErrorCode == azblob.StorageErrorCodeBlobNotFound {
+			return true
+		}
+	}
+	return false
+}
+
+// buildServiceClient attempts to build an azblob.ServiceClient with
+// authentication configured based on the data from the Secret.
+//
+// It configures (in order):
+// - azidentity.ManagedIdentityCredential for a Resource ID, when a
+//   "resourceId" entry is found.
+// - azidentity.ManagedIdentityCredential for a User ID, when a "clientId"
+//   but no "tenantId" entry is found.
+// - azidentity.ClientSecretCredential when a "tenantId", "clientId" and
+//   "clientSecret" is found.
+// - azblob.SharedKeyCredential when a "accountName" and "accountKey" are
+//   found.
+// - Client without credentials.
+func buildServiceClient(obj *sourcev1.Bucket, secret *corev1.Secret) (_ azblob.ServiceClient, err error) {
+	// Without a Secret, we can return a simple client.
+	if secret == nil || len(secret.Data) == 0 {
+		return azblob.NewServiceClientWithNoCredential(obj.Spec.Endpoint, nil)
+	}
+
+	// Attempt AAD Token Credential options first.
+	var token azcore.TokenCredential
+	if resourceID, ok := secret.Data[ResourceIDField]; ok {
+		if token, err = azidentity.NewManagedIdentityCredential(&azidentity.ManagedIdentityCredentialOptions{
+			ID: azidentity.ResourceID(resourceID)},
+		); err != nil {
+			return
+		}
+	}
+	if clientID, hasClientID := secret.Data[ClientIDField]; hasClientID {
+		tenantID, hasTenantID := secret.Data[TenantIDField]
+		if !hasTenantID {
+			if token, err = azidentity.NewManagedIdentityCredential(&azidentity.ManagedIdentityCredentialOptions{
+				ID: azidentity.ClientID(clientID),
+			}); err != nil {
+				return
+			}
+		}
+		if clientSecret, hasClientSecret := secret.Data[ClientSecretField]; hasClientSecret {
+			if token, err = azidentity.NewClientSecretCredential(string(tenantID), string(clientID), string(clientSecret), nil); err != nil {
+				return
+			}
+		}
+	}
+	if token != nil {
+		return azblob.NewServiceClient(obj.Spec.Endpoint, token, nil)
+	}
+
+	// Fallback to Shared Key Credential.
+	if accountName, hasAccountName := secret.Data[AccountNameField]; hasAccountName {
```

Comment:
I don't think we need the user to provide the account name since it can be parsed from the endpoint.

```go
endpointURL, err = url.Parse(obj.Spec.Endpoint)
if err != nil {
   return
}
accountName := strings.Split(endpointURL.Host, ".")[0]

Reply:
Thanks for having a look! I had thought about this as well (and noticed it in your version), but wondered if this is absolutely true in all Azure environments out there (and e.g. having it in another part of the URL, or some format that trips the simple split is deemed impossible).

Reply:
Currently it's absolutely true in Azure environments when the endpoint is formed in `https://accountName.storageAccountSuffix` and only digits/letters are allowed in `accountName`.
Fine for me to ask for the `accountName` in the secret for some flexibility going forwards.